From edc8feb4bd81f64a8cbe44bb8cb5e5d8adb1ba64 Mon Sep 17 00:00:00 2001
From: Robert Dabrowski <robertx.dabrowski@intel.com>
Date: Thu, 13 Feb 2020 08:54:47 +0100
Subject: [PATCH] PONRTSYS-6674: Deregister dev notify event passes
 uninitialized data to dp_event_cb - fix register/deregister event callback -
 fix removing dev from counters update list

---
 drivers/net/datapath/mib_counters/dp_mib_counters.c | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/drivers/net/datapath/mib_counters/dp_mib_counters.c b/drivers/net/datapath/mib_counters/dp_mib_counters.c
index 63925996c07a..9226ee514069 100644
--- a/drivers/net/datapath/mib_counters/dp_mib_counters.c
+++ b/drivers/net/datapath/mib_counters/dp_mib_counters.c
@@ -189,7 +189,7 @@ s32 rm_dev_from_list(struct net_device *dev, s32 port_id, s32 sub_if, s32 inst)
 {
 	struct pmac_port_info *port_info;
 	struct mib_counters *mib_cnt;
-	struct list_head *p;
+	struct list_head *p,*q;
 	s32 sub_if_id_grp;
 
 	/* TODO: tune the range check */
@@ -215,7 +215,7 @@ s32 rm_dev_from_list(struct net_device *dev, s32 port_id, s32 sub_if, s32 inst)
 		 sub_if_id_grp);
 
 	write_lock(&dev_list_lock);
-	list_for_each(p, &registered_devices_list)
+	list_for_each_safe(p, q, &registered_devices_list)
 	{
 		mib_cnt = list_entry(p, struct mib_counters, dev_list);
 
@@ -226,7 +226,7 @@ s32 rm_dev_from_list(struct net_device *dev, s32 port_id, s32 sub_if, s32 inst)
 		if ((mib_cnt->port_id == port_id) &&
 		    (mib_cnt->sub_if_id_grp == sub_if_id_grp) &&
 		    (mib_cnt->inst == inst)) {
-			DP_DEBUG(DP_DBG_FLAG_REG, "MIB CNT - removed device: %s port_id = %d, sub_if_id_grp = %d\n",
+			DP_DEBUG(DP_DBG_FLAG_REG, "MIB CNT - removing device: %s port_id = %d, sub_if_id_grp = %d\n",
 				 dev->name, port_id, sub_if_id_grp);
 			list_del(p);
 			devm_kfree(&g_dp_dev->dev, mib_cnt);
@@ -428,7 +428,7 @@ static int __init mib_counters_mod_init(void)
 	/* Registering of net dev register callback */
 	event_info.inst = 0;
 	event_info.owner = DP_EVENT_OWNER_MIB;
-	event_info.type = DP_EVENT_REGISTER_SUBIF;
+	event_info.type = DP_EVENT_REGISTER_SUBIF | DP_EVENT_DE_REGISTER_SUBIF;
 	event_info.id = 0;
 	event_info.dp_event_cb = subif_register_cb;
 
@@ -453,13 +453,13 @@ static void __exit mib_counters_mod_exit(void)
 	/* Deregistering the callback function */
 	event_info.inst = 0;
 	event_info.owner = DP_EVENT_OWNER_MIB;
-	event_info.type = DP_EVENT_DE_REGISTER_SUBIF;
+	event_info.type = DP_EVENT_REGISTER_SUBIF | DP_EVENT_DE_REGISTER_SUBIF;
 	event_info.id = g_notify_cb_id;
 	event_info.dp_event_cb = subif_register_cb;
 
 	ret = dp_register_event_cb(&event_info, DP_F_DEREGISTER);
 	if (ret != DP_SUCCESS) {
-		pr_err("Can't deregister DP_EVENT_REGISTER_SUBIF callback\n");
+		pr_err("Can't deregister SUBIF notify callback\n");
 		return;
 	}
 
